In [1]:
import pandas as pd
import numpy as np
import csv
from scipy import stats

In [2]:
df= pd.read_csv('StudentsPerformance_mod.csv')
df.head(10)

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college           NaN   
5  female        group B          associate's degree      standard   
6     NaN        group B                         NaN      standard   
7    male            NaN                some college  free/reduced   
8    male        group D                 high school  free/reduced   
9  female        group B                 high school  free/reduced   

  test preparation course  math score  reading score  writing score  
0                    none        72.0           72.0           74.0  
1               completed        69.0           90.0           88.0  
2                    none        90.0           95.0           93.0  
3                    none        47.0           57.0           44.0  
4                    none        76.0           78.0           75.0  
5                    none        71.0           83.0           78.0  
6               completed         NaN           95.0           92.0  
7                    none        40.0           43.0           39.0  
8               completed        64.0           64.0           67.0  
9                    none        38.0           60.0           50.0

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   gender                       992 non-null    object 
 1   race/ethnicity               988 non-null    object 
 2   parental level of education  992 non-null    object 
 3   lunch                        990 non-null    object 
 4   test preparation course      996 non-null    object 
 5   math score                   989 non-null    float64
 6   reading score                997 non-null    float64
 7   writing score                995 non-null    float64
dtypes: float64(3), object(5)
memory usage: 62.6+ KB


In [4]:
df.describe()

math score  reading score  writing score
count   989.000000     997.000000     995.000000
mean     66.984934      69.253762      69.011055
std      37.535832      20.665986      33.911544
min    -200.000000     -50.000000    -100.000000
25%      57.000000      59.000000      58.000000
50%      66.000000      70.000000      69.000000
75%      77.000000      79.000000      79.000000
max    1005.000000     500.000000    1000.000000

I will use the mode (most frequent data values) to fill the NA values, as the gender, race/ethnicity, parental level of education, lunch, test preparation course data values are categorical ones.

In [5]:
np.sum(df['gender'].isnull())

8

In [6]:
np.mean(df['gender'].isnull())*100

0.8

In [7]:
mod= stats.mode(df['gender'])[0]
print(mod)

['female']


In [8]:
df['gender'].fillna(stats.mode(df['gender']).mode[0], inplace=True)

In [9]:
np.sum(df['race/ethnicity'].isnull())

12

In [10]:
np.mean(df['race/ethnicity'].isnull())*100

1.2

In [11]:
mod= stats.mode(df['race/ethnicity'])[0]
print(mod)

['group C']


In [12]:
df['race/ethnicity'].fillna(stats.mode(df['race/ethnicity']).mode[0], inplace=True)

In [13]:
np.sum(df['parental level of education'].isnull())

8

In [14]:
np.mean(df['parental level of education'].isnull())*100

0.8

In [15]:
mod= stats.mode(df['parental level of education'])[0]
print(mod)

['some college']


In [16]:
df['parental level of education'].fillna(stats.mode(df['parental level of education']).mode[0], inplace=True)

In [17]:
np.sum(df['lunch'].isnull())

10

In [18]:
np.mean(df['lunch'].isnull())*100

1.0

In [19]:
mod= stats.mode(df['lunch'])[0]
print(mod)

['standard']


In [20]:
df['lunch'].fillna(stats.mode(df['lunch']).mode[0], inplace=True)

In [21]:
np.sum(df['test preparation course'].isnull())

4

In [22]:
np.mean(df['test preparation course'].isnull())*100

0.4

In [23]:
mod= stats.mode(df['test preparation course'])[0]
print(mod)

['none']


In [24]:
df['test preparation course'].fillna(stats.mode(df['test preparation course']).mode[0], inplace=True)

I will use quartiles to check for outliers, as when distribution of data values is skewed to the left,  mean will be pulled towards the left (same thing applies to the right skewed distribution). Therefore, mean is not an accurate check for outliers.

Furthermore, I will use the median (50th percentile) to fill the NA values, as the math score, reading score, writing score data values are numerical ones.

In [25]:
np.sum(df['math score'].isnull())

11

In [26]:
np.mean(df['math score'].isnull())*100

1.0999999999999999

In [27]:
df['math score'].quantile(0.5)

66.0

In [28]:
df['math score'].fillna(df['math score'].quantile(0.5), inplace=True)

In [29]:
#checking for low outliers in Math score column
df['math score'].quantile(0.25)-(1.5*(df['math score'].quantile(0.75)-df['math score'].quantile(0.25)))

27.0

There are data values in the Math score column below 27
Therefore, there are low outliers

In [30]:
#checking for high outliers in Math score column
df['math score'].quantile(0.75)+(1.5*(df['math score'].quantile(0.75)-df['math score'].quantile(0.25)))

107.0

There are data values in the Math score column above 107
Therefore, there are high outliers

In [31]:
np.sum(df['reading score'].isnull())

3

In [32]:
np.mean(df['reading score'].isnull())*100

0.3

In [33]:
df['reading score'].quantile(0.5)

70.0

In [34]:
df['reading score'].fillna(df['reading score'].quantile(0.5), inplace=True)

In [35]:
#checking for low outliers in Reading score column
df['reading score'].quantile(0.25)-(1.5*(df['reading score'].quantile(0.75)-df['reading score'].quantile(0.25)))

29.0

There are data values in the Reading score column below 29
Therefore, there are low outliers

In [36]:
#checking for high outliers in Reading score column
df['reading score'].quantile(0.75)+(1.5*(df['reading score'].quantile(0.75)-df['reading score'].quantile(0.25)))

109.0

There are data values in the Reading score column above 109
Therefore, there are high outliers

In [37]:
np.sum(df['writing score'].isnull())

5

In [38]:
np.mean(df['writing score'].isnull())*100

0.5

In [39]:
df['writing score'].quantile(0.5)

69.0

In [40]:
df['writing score'].fillna(df['writing score'].quantile(0.5), inplace=True)

In [41]:
#checking for low outliers in writing score column
df['writing score'].quantile(0.25)-(1.5*(df['writing score'].quantile(0.75)-df['writing score'].quantile(0.25)))

26.5

There are data values in the Writing score column below 26.5
Therefore, there are low outliers

In [42]:
#checking for low outliers in writing score column
df['writing score'].quantile(0.75)+(1.5*(df['writing score'].quantile(0.75)-df['writing score'].quantile(0.25)))

110.5

There are data values in the Writing score column above 110.5
Therefore, there are high outliers

In [43]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   gender                       1000 non-null   object 
 1   race/ethnicity               1000 non-null   object 
 2   parental level of education  1000 non-null   object 
 3   lunch                        1000 non-null   object 
 4   test preparation course      1000 non-null   object 
 5   math score                   1000 non-null   float64
 6   reading score                1000 non-null   float64
 7   writing score                1000 non-null   float64
dtypes: float64(3), object(5)
memory usage: 62.6+ KB


In [44]:
df.head(10)

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   
5  female        group B          associate's degree      standard   
6  female        group B                some college      standard   
7    male        group C                some college  free/reduced   
8    male        group D                 high school  free/reduced   
9  female        group B                 high school  free/reduced   

  test preparation course  math score  reading score  writing score  
0                    none        72.0           72.0           74.0  
1               completed        69.0           90.0           88.0  
2                    none        90.0           95.0           93.0  
3                    none        47.0           57.0           44.0  
4                    none        76.0           78.0           75.0  
5                    none        71.0           83.0           78.0  
6               completed        66.0           95.0           92.0  
7                    none        40.0           43.0           39.0  
8               completed        64.0           64.0           67.0  
9                    none        38.0           60.0           50.0